### PR TITLE
Add multi-output-arg version of size()

### PR DESCRIPTION
**Pull request by rdzman**
Thanks for your work on table support for Octave. Hoping it will be completed and included in Octave core soon.

### CODE REVIEW

**Comment by apjanke** [comment]
> Thanks for your work on table support for Octave.

Thank you!

> Hoping it will be completed and included in Octave core soon.

Discussions with the core Octave developers about this are happening as we speak! However, this is kind of a large and complex project. "Included in Octave core", yes! "soon", don't get your hopes up!

Thanks for the PR here; this is something I totally overlooked. There's a couple fiddly details to work out; see the code review I'll be posting soon.

File: inst/table.m [hunk 1]
```diff
@@ -424,19 +424,26 @@ function prettyprint (this)
     ## -*- texinfo -*-
     ## @node table.size
     ## @deftypefn {Method} {@var{sz} =} size (@var{obj})
+    ## @deftypefnx {Method} {[@var{nr}, @var{nv}] =} size (@var{obj})
```
Comment (apjanke) [inline inst/table.m:427]: Need another line here for the N-argout case that indicates you can have _any_ number of argouts.

File: inst/table.m [hunk 2]
```diff
@@ -424,19 +424,26 @@ function prettyprint (this)
...
     ## Gets the size of a table.
     ##
     ## For tables, the size is [number-of-rows x number-of-variables].
     ## This is the same as @code{[height(obj), width(obj)]}.
     ##
     ## @end deftypefn
-    function out = size (this, dim)
-      if nargin == 1
-        out = [height(this), width(this)];
+    function [out, nv] = size (this, dim)
```
Comment (apjanke) [inline inst/table.m:435]: I think this needs to become `varargout = ...` instead of a fixed list of argouts.

File: inst/table.m [hunk 3]
```diff
@@ -424,19 +424,26 @@ function prettyprint (this)
...
+      if nargout == 1
```
Comment (apjanke) [inline inst/table.m:436]: So, here's the deal: instead of just handling the `nargout ==1` and `nargout == 2` cases, I think we need to handle _any_ number of argouts.

Zero argouts should have the same behavior as `nargout == 1`. The `if nargout == 1 ... else` logic here is going to toss `nargout == 0` into the `nargout == 2` bin and give you just the height, I think. That's wrong; you should get the full 2-D size in that case.

And we need to extend this to handle nargout > 2: in Octave, arrays are actually many- or infinite-dimensioned, and a 2-D array like a table array actually has an infinite number of trailing singleton dimensions whose size is all 1. So if I do this:

```octave
x = [1; 2; 3];
y = [420; 69; 42];
tbl = table(x, y);
[d1, d2, d3, d4, d5, d6] = size(tbl)
```

I think that should actually work, and you should get:
* d1 = 3
* d2 = 2
* d3, d4, d5, d6 all = 1

Basically, I think we need to extend this logic to take any number of argouts, and when nargout > 2, fill all the nargout{i>2} with 1s.

And in the 2-argin case, whenever dim is > 2, return 1.

What do you think?

File: inst/table.m [hunk 4]
```diff
@@ -424,19 +424,26 @@ function prettyprint (this)
...
+        elseif dim == 2
+          out = width(this);
+        else
+          out = height(this);
+        end
       else
-        sz = size (this);
-        out = sz(dim);
+        out = height(this);
+        nv  = width(this);
```
Comment (apjanke) [inline inst/table.m:446]: Got an extra space here. Tablicious and the standard GNU Octave style generally do not do equal-sign alignment.

**Review by apjanke — CHANGES_REQUESTED** [review]

**Comment by apjanke** [comment]
Because this fixes a significant flaw in the behavior of `table.size`, which is a fundamental operation, I am scheduling this for the next patch release milestone.

File: inst/table.m (hunk 1 above)
Comment (rdzman) [reply inst/table.m:427]: Isn't a `table` by definition only 2-D ... rows x variables? Do we really need to support the N-argout case?

Comment (apjanke) [reply inst/table.m:427]: Not quite: By definition, `table`s only have _nonzero_ length along dimensions 1 and 2, since they are 2-D arrays. And when you ask for `size(tbl)` you'll get a 2-long array back. And `ndims(tbl)` will always say "2".

But in Matlab and Octave, arrays of any dimensionality also, by definition, have an implicit number of 1-long trailing dimensions, which are only displayed or reported when you ask about them in certain ways, but nonetheless exist.

Consider the case of a scalar numeric:

```
>> x = 420.69;
>> size(x)
ans =
   1   1

>> size(x,3)
ans = 1
>> size(x,420)
ans = 1
>> [a,b,c,d,e,f,g,h] = size(x)
a = 1
b = 1
c = 1
d = 1
e = 1
f = 1
g = 1
h = 1
>>
```

Don't look at me, man; I don't make the rules, I just follow them. ;)

File: inst/table.m (hunk 3 above)
Comment (rdzman) [reply inst/table.m:436]: Ok, I agree that the correct behavior for a nargout > 2 is 1's for the additional dimensions (not an error as we have now). I'll try to fix it tomorrow.

Comment (apjanke) [reply inst/table.m:436]: Cool beans. No rush – it's the weekend, and this is a volunteer project. Get to it when you get to it. I'll wait on your PR revision before making any other moves here.

File: inst/table.m (hunk 4 above)
Comment (apjanke) [reply inst/table.m:446]: Cool beans! No rush - it's the weekend, and this is a volunteer project. I'll wait on your PR revision before making any other moves here.

**Comment by rdzman** [comment]
Ok, this update should handle an arbitrary number of output args. And `nargout == 0` behaves like `nargout == 1`. Not sure if it needs to throw errors for values of `dim` that are not positive integers, but I suppose that's a separate issue.

**Comment by apjanke** [comment]
Yes, this is looking good to me now. I'll do some testing on it and then get this merged.

Thanks!

**Comment by apjanke** [comment]
I have reviewed the code and tested this interactively, and it looks good to me. If I were a better programmer, I'd write some unit tests for this, but who has the time?

Merged.

Thank you!
